src/query/functions/aggregation/base_async_benchmark_test.go

```diff
@@ -0,0 +1,242 @@
+	require.NoError(b, err)
+
+	asyncOp, valid := op.(baseOp)
+	require.True(b, valid)
+	return asyncOp.AsyncNode(c, workerPool), sink
+}
+
+func benchWorkerPools(b *testing.B) xsync.PooledWorkerPool {
+	poolSize := 16
+	poolOpts := xsync.NewPooledWorkerPoolOptions().
+		SetGrowOnDemand(true).
+		SetNumShards(int64(poolSize))
+	workerPool, err := xsync.NewPooledWorkerPool(poolSize, poolOpts)
+	require.NoError(b, err)
+	workerPool.Init()
+	return workerPool
+}
+
+type mockBlock struct {
+	seriesCount int
+	stepCount   int
+}
+
+func (b *mockBlock) Close() error { return nil }
+func (b *mockBlock) AsyncStepIter() (block.AsyncStepIter, error) {
+	metas := make([]block.SeriesMeta, b.seriesCount)
+	for i := 0; i < b.seriesCount; i++ {
+		b := []byte(fmt.Sprint(i))
+		metas[i] = block.SeriesMeta{
+			Name: "a",
+			Tags: models.NewTags(1, models.NewTagOptions()).AddTag(
+				models.Tag{
+					Name:  b,
+					Value: b,
+				},
+			),
+		}
+	}
+
+	return &mockIter{
+		stepCount:   b.stepCount,
+		seriesCount: b.seriesCount,
+		metas:       metas,
+	}, nil
+}
+
+type mockIter struct {
+	seriesCount int
+	stepCount   int
+	step        int
+	metas       []block.SeriesMeta
+}
+
+func (it *mockIter) SeriesMeta() []block.SeriesMeta {
+	return it.metas
+}
+
+func (it *mockIter) Meta() block.Metadata {
+	return block.Metadata{}
+}
+
+func (it *mockIter) StepCount() int {
+	return it.stepCount
+}
+
+func (it *mockIter) Next() bool {
+	it.step = it.step + 1
+	return it.step < it.stepCount+1
+}
+
+func (it *mockIter) Current() <-chan block.Step {
+	ch := make(chan block.Step)
+	go func() {
+		ch <- &mockStep{
+			seriesCount: it.seriesCount,
+		}
+	}()
+	return ch
+}
+
+func (it *mockIter) ValuesChannel() <-chan block.IndexedStep {
+	ch := make(chan block.IndexedStep, it.stepCount)
+	for i := 0; i < it.stepCount; i++ {
+		ch <- block.IndexedStep{
+			Idx: i,
+			Step: &mockStep{
+				seriesCount: it.seriesCount,
+			},
+		}
+	}
+	close(ch)
+	return ch
+}
+
+func (it *mockIter) Err() error { return nil }
+func (it *mockIter) Close()     {}
+
+type mockStep struct {
+	seriesCount int
+}
+
+func (s *mockStep) Time() time.Time {
+	return time.Time{}
+}
+
+func (s *mockStep) Values() []float64 {
+	values := make([]float64, s.seriesCount)
+	ts.Memset(values, float64(12))
+	return values
+}
+
+func BenchmarkSteps1Series100Step(b *testing.B) {
+	benchmarkSteps(b, 1, 100)
+}
+
+func BenchmarkValueChannel1Series100Step(b *testing.B) {
+	benchmarkValueChannel(b, 1, 100)
+}
+
+func BenchmarkSteps100Series100Step(b *testing.B) {
+	benchmarkSteps(b, 100, 100)
+}
+
+func BenchmarkValueChannel100Series100Step(b *testing.B) {
+	benchmarkValueChannel(b, 100, 100)
+}
+
+func BenchmarkSteps10kSeries100Step(b *testing.B) {
+	benchmarkSteps(b, 10000, 100)
+}
+
+func BenchmarkValueChannel10kSeries100Step(b *testing.B) {
+	benchmarkValueChannel(b, 10000, 100)
+}
+
+func BenchmarkSteps100Series10kStep(b *testing.B) {
+	benchmarkSteps(b, 100, 10000)
```

Comment:
Not that it matters too much, but this could read nicely as a table driven benchmark:

```
func BenchmarkValueChannel(b *testing.B) {
   cases := []struct{
          Series int
          Step int
   }{
      {100, 100},
      {10000, 100},
      ...
    }

     for _, tc := range cases {
           b.Run("...", func(b *testing.B) {
                	benchmarkSteps(b, tc.Series, tc.Steps)
           })
     }
}
```

Reply:
Didn't know we could apply table driven tests to benchmarks but I guess it makes sense